keps/sig-scheduling/4671-gang-scheduling/README.md

```diff
@@ -469,19 +519,269 @@ states (e.g. not yet block resources) will help with improving the scheduling ac
+
+Alternative 2 (Store a gang representative):
+
+Only one "representative" Pod from the gang is allowed in the `activeQ` at a time.
+Others are held in a separate internal structure (e.g., a new map inside the queue).
+When the representative is popped, it pulls the rest of the gang for the Workload Cycle.
+* *Pros:* Makes it easier to obtain all pod group's pods, reduces queue size.
+* *Cons:* High complexity in managing the lifecycle of the representative
+  (e.g., what if the representative Pod is deleted or other changes to the workload happen?
+  Would need a workload manager to handle all such cases).
+
+Alternative 3 (Dedicated PodGroup queue):
+
+Introduce a completely separate queue for PodGroups alongside the `activeQ` for Pods.
+The scheduler would pop the item (Pod or PodGroup) with the highest priority/earliest timestamp.
+Pods belonging to an enqueued PodGroup won't be allowed in the `activeQ`.
+* *Pros:* Clean separation of concerns. Can easily use the Workload scheduling priority.
+  Can report dedicated logs and metrics with less confusion to the user.
+* *Cons:* Significant and non-trivial architectural change to the scheduling queue
+  and `scheduleOne` loop.
+<<[/UNRESOLVED]>>
+```
+
+#### Scheduling Algorithm
+
+The internal algorithm for placing the group utilizes the optimization defined
+in *Opportunistic Batching* ([KEP-5598](https://kep.k8s.io/5598)) for improved performance.
+The approach described below allows mitigating some restrictions of that feature, e.g.,
+by sorting the Pods appropriately by their signatures. In case Opportunistic Batching
+is disabled or not applicable, this falls back to non-optimized filtering and scoring for each Pod.
+
+1. The scheduler iterates through the retrieved Pods and groups
+   them into homogeneous sub-groups (using the signatures defined in
+   [KEP-5598](https://kep.k8s.io/5598)).
+
+2. These sub-groups are sorted. Initially, we sort by the highest priority
+   of the sub-group (assuming homogeneity enforces uniform sub-group priority).
+   In the future, sorting may use the size of the sub-group (larger groups first) to
+   tackle the hardest placement problems early.
+
+3. The scheduler iterates through the sorted sub-groups. It finds a feasible node
+   for each pod from a sub-group using standard filtering and scoring phases.
+   It also utilizes the Opportunistic Batching feature where possible,
+   reducing overall scheduling time.
+
+   * If a pod fits, it is tentatively assigned.
+   * If a pod cannot fit, the scheduler tries preemption by running
+     the `PostFilter` extension point. *Note:* With workload-aware preemption
+     this phase will be replaced by a workload-level algorithm.
+     * If preemption is successful, the pod is nominated on the selected node.
+     * If preemption fails, the pod is considered unscheduled for this cycle.
+
+   The phase can effectively stop once `minCount` pods have a placement,
```

Comment:
From the POV of optimizing locally, we should continue scheduling all pod (and not stop after minCount) - what changes is that the unschedulability of further pods shouldn't make the whole group unschedulable.

Given that we have `minCount` at the whole PodGroup level, the local optimization is actually what we should probably do anyway (it would be a different story if we would have a separate minCount per signature, but it's not the case).